### PR TITLE
Vessel history/vh 82

### PR DESCRIPTION
**Pull request by fertandil87**
image
image

TICKETS INCLUDED:
https://globalfishingwatch.atlassian.net/browse/VH-85
https://globalfishingwatch.atlassian.net/browse/VH-60
https://globalfishingwatch.atlassian.net/browse/VH-56
https://globalfishingwatch.atlassian.net/browse/VH-82
https://globalfishingwatch.atlassian.net/browse/VH-53

### CODE REVIEW

File: applications/vessel-history/src/features/vessels/vessels.thunks.ts [hunk 1]
```diff
@@ -1,30 +1,64 @@
...
-  const id = selectVessel(state)
-  if (!id || id === 'NA') return null
+  const vesselID = selectVesselId(state)
+  const tmtID = selectTmtId(state)
+  const dataset = selectDataset(state)
+  const id = selectVesselProfileId(state)
+  const data = await fetchVesselInfo(vesselID, tmtID, dataset)
+  const vessels = selectVessels(state)
+
+  if (vessels[id]) return
```
Comment (rdgfuentes) [inline applications/vessel-history/src/features/vessels/vessels.thunks.ts:60]: I don't understand the purpose of this and why it's done after fetchVesselInfo in the precious line.

**Review by rdgfuentes — CHANGES_REQUESTED** [review]
Look at my comments below and also take into account:
* Add some tests to cover the most relevant parts of this feature at least
* Fix the tests that are not passing 
```bash
╰─$ yarn test:vessel-history                                                                                                                                                                                        1 ↵
yarn run v1.19.1
$ cross-env lerna run test --scope @globalfishingwatchapp/vessel-history
lerna notice cli v3.22.1
lerna info versioning independent
lerna notice filter including "@globalfishingwatchapp/vessel-history"
lerna info filter [ '@globalfishingwatchapp/vessel-history' ]
lerna info Executing command in 1 package: "yarn run test"
@globalfishingwatchapp/vessel-history: $ CI=true react-scripts test --passWithNoTests
@globalfishingwatchapp/vessel-history: PASS src/utils/vessel.test.tsx
@globalfishingwatchapp/vessel-history: PASS src/components/Splash.test.tsx
@globalfishingwatchapp/vessel-history: PASS src/features/search/SearchPlaceholders.test.tsx
@globalfishingwatchapp/vessel-history: PASS src/App.test.tsx
@globalfishingwatchapp/vessel-history: FAIL src/features/vessel-list-item/VesselListItem.test.tsx (5.253 s)
@globalfishingwatchapp/vessel-history:   ● Console
@globalfishingwatchapp/vessel-history:     console.error
@globalfishingwatchapp/vessel-history:       Error: Uncaught [Error: Could not find "store" in the context of "Connect(Link)". Either wrap the root component in a <Provider>, or pass a custom React context provider to <Provider> and the corresponding React context consumer to Connect(Link) in connect options.]
...
```
* Look at the warnings in the console and fix those on the files you modified in those PR so that we get rid of them progressively
image

File: applications/vessel-history/src/features/vessels/vessels.thunks.ts (hunk 1 above)
Comment (j8seangel) [reply applications/vessel-history/src/features/vessels/vessels.thunks.ts:60]: I guess... if you have already requested the vessel info you can avoid that request and use the "cached" data

File: applications/vessel-history/src/classes/vessel.class.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+export interface InfoValue {
+  data: string | null
+  historic: Array<HistoricValue>
+}
+export interface VesselInfoValue {
+  value: InfoValue | null
+  source: VesselAPISource[]
+}
+
+export class VesselInfo {
```
[the remark this answers is not in the document]
Comment (j8seangel) [reply applications/vessel-history/src/classes/vessel.class.ts:20]: I'd suggest to implement and extend the existing in [api-types](https://github.com/GlobalFishingWatch/frontend/blob/vessel-history/VH-82/packages/api-types/src/vessel.ts#L1) package so we have the same types in every app

**Comment by j8seangel** [comment]
> * Look at the warnings in the console and fix those on the files you modified in those PR so that we get rid of them progressively
>   image

I've updated the vscode config to show warnings in the editor: https://github.com/GlobalFishingWatch/frontend/commit/8f4495b161d7e7c54186c15d6e1e27e89b2a1089

File: applications/vessel-history/src/features/profile/components/Info.tsx [hunk 3]
```diff
@@ -40,64 +21,33 @@ const Info: React.FC<InfoProps> = (props): React.ReactElement => {
...
-            </div>
-            <div>
-              <label>OWNER</label>
-              <span>NENG DE OCEAN COMPANY LTD</span>
-            </div>
-            <div>
-              <label>OPERATOR</label>
-              <span>NENG DE OCEAN COMPANY LTD</span>
-            </div>
+            <InfoField label={'Name'} field={vessel.getName()}></InfoField>
```
[the remark this answers is not in the document]
Comment (j8seangel) [reply applications/vessel-history/src/features/profile/components/Info.tsx:24]: I agree with @rdgfuentes, I like the idea of having a class to manage the internal logic but I'm not sure if it should reach the component level, I think it is simpler if we thread as a plain object.
So... what about having a method that invokes all of the internal ones and return a plain object with all the properties like:
```jsx
export class VesselInfo {
  read() || .get() || any other better name {
    return {
      name: this.getName(),
      type: this.getType()
    }
  }
}
```

then populated in the vessels.selector ?

```jsx
return new VesselInfo(vessels[id].gfwData, vessels[id].tmtData).theMethod()
```

and finally in the component (components/Info.tsx) we just use 
```jsx
InfoField label={'Name'} field={vessel.name}></InfoField>
```

**Comment by j8seangel** [comment]
BTH the tab `)` typo was fixed in the latest version of the ui-components, please upgrade it

File: applications/vessel-history/src/features/profile/Profile.tsx [hunk 4]
```diff
@@ -1,33 +1,62 @@
...
-        )}
-        {selectedTab === 2 && (
-          <RecentActivity vesselID={vesselID} setLastPortVisit={setLastPortVisit} />
+      <header className={styles.header}>
+        <IconButton
+          type="border"
+          size="default"
+          icon="arrow-left"
+          className={styles.backButton}
+          onClick={async () => {
```
[the remark this answers is not in the document]
Comment (rdgfuentes) [reply applications/vessel-history/src/features/profile/Profile.tsx:41]: Fixed on https://github.com/GlobalFishingWatch/frontend/pull/349

File: applications/vessel-history/src/features/vessel-list-item/VesselListItem.tsx [hunk 5]
```diff
@@ -29,7 +30,9 @@ const VesselListItem: React.FC<ListItemProps> = (props): React.ReactElement => {
           className={styles.deleteSaved}
         ></IconButton>
       )}
-      <h3>{vessel?.shipname}</h3>
+      <Link to={['profile', vessel.dataset ?? 'NA', vessel.id ?? 'NA', vessel.vesselMatchId]}>
```
[the remark this answers is not in the document]
Comment (rdgfuentes) [reply applications/vessel-history/src/features/vessel-list-item/VesselListItem.tsx:33]: fixed on https://github.com/GlobalFishingWatch/frontend/pull/349

File: applications/vessel-history/src/classes/vessel.class.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+          data: historicValue.value,
+          start: historicValue.firstSeen ? new Date(historicValue.firstSeen) : null,
+          end: historicValue.endDate ? new Date(historicValue.endDate) : null,
+        }
+        return value
+      }) || []
+    if (gfwValue) {
+      dataValues.push({
+        data: gfwValue,
+        start: gfwData?.firstTransmissionDate ? new Date(gfwData?.firstTransmissionDate) : null,
```
[the remark this answers is not in the document]
Comment (rdgfuentes) [reply applications/vessel-history/src/classes/vessel.class.ts:54]: fixed in https://github.com/GlobalFishingWatch/frontend/pull/349

File: applications/vessel-history/src/classes/vessel.class.ts [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        data: null,
+        historic: [],
+      }
+    }
+    const dataValues =
+      tmtValue?.map((historicValue: AnyValueList) => {
+        console.log(historicValue.firstSeen)
+        const value: HistoricValue = {
+          data: historicValue.value,
+          start: historicValue.firstSeen ? new Date(historicValue.firstSeen) : null,
```
[the remark this answers is not in the document]
Comment (rdgfuentes) [reply applications/vessel-history/src/classes/vessel.class.ts:46]: fixed in https://github.com/GlobalFishingWatch/frontend/pull/349

File: applications/vessel-history/src/classes/vessel.class.ts [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  ): InfoValue {
+    if ((!tmtValue || !tmtValue.length) && !gfwValue) {
+      return {
+        data: null,
+        historic: [],
+      }
+    }
+    const dataValues =
+      tmtValue?.map((historicValue: AnyValueList) => {
+        console.log(historicValue.firstSeen)
```
[the remark this answers is not in the document]
Comment (rdgfuentes) [reply applications/vessel-history/src/classes/vessel.class.ts:43]: fixed in https://github.com/GlobalFishingWatch/frontend/pull/349

File: applications/vessel-history/src/classes/vessel.class.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+      }) || []
+    if (gfwValue) {
+      dataValues.push({
+        data: gfwValue,
+        start: gfwData?.firstTransmissionDate ? new Date(gfwData?.firstTransmissionDate) : null,
+        end: gfwData?.lastTransmissionDate ? new Date(gfwData?.lastTransmissionDate) : null,
+      })
+    }
+    if (gfwHistoricValue && gfwHistoricValue.length) {
+      console.log(gfwHistoricValue)
```
[the remark this answers is not in the document]
Comment (rdgfuentes) [reply applications/vessel-history/src/classes/vessel.class.ts:59]: fixed in https://github.com/GlobalFishingWatch/frontend/pull/349

File: applications/vessel-history/src/features/profile/components/InfoField.tsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  if (!field.value?.data) {
+    return <div></div>
+  }
+
+  return (
+    <div className={styles.identifierField}>
+      <label>{props.label}</label>
+      <span onClick={() => displachOpenModal(field)}>
+        {field.value?.data}
+        {field.value?.historic.length && <Icon icon="search"></Icon>}
```
[the remark this answers is not in the document]
Comment (rdgfuentes) [reply applications/vessel-history/src/features/profile/components/InfoField.tsx:29]: fixed in #349

File: applications/vessel-history/src/features/profile/components/Info.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (rdgfuentes) [reply applications/vessel-history/src/features/profile/components/Info.tsx:24]: To be addressed in another branch `vessel-history/VH-82-refactor-thunks`

**Comment by rdgfuentes** [comment]
Refactors around `vessel.class.ts` and how we are getting the vessel data from different sources and combining it will be addressed in another PR ( branch `vessel-history/VH-82-refactor-thunks`)

**Review by rdgfuentes — APPROVED** [review]
Refactors around vessel.class.ts and how we are getting the vessel data from different sources and combining it will be addressed in another PR ( branch vessel-history/VH-82-refactor-thunks)

**Pull request merged**